### PR TITLE
[opencypher compatibility] Support using exists with dynamic map

### PR DESCRIPTION
**Pull request by Aiee**
As title.

Dependency of https://github.com/vesoft-inc/nebula-graph/pull/973

### CODE REVIEW

File: src/common/expression/PredicateExpression.cpp [hunk 1]
```diff
@@ -17,10 +17,18 @@ std::unordered_map<std::string, PredicateExpression::Type> PredicateExpression::
...
-    auto* attributeExpr = static_cast<AttributeExpression* >(collection_.get());
-    auto& container = attributeExpr->left()->eval(ctx);
-    auto& key = attributeExpr->right()->eval(ctx);
+    DCHECK(collection_->kind() == Expression::Kind::kAttribute ||
+           collection_->kind() == Expression::Kind::kSubscript);
+
+    BinaryExpression* containerExpr;
+    if (collection_->kind() == Expression::Kind::kAttribute) {
+        containerExpr = static_cast<AttributeExpression*>(collection_.get());
+    } else {
```
Comment (Shylock-Hg) [inline src/common/expression/PredicateExpression.cpp:26]: else if ( == kSubscript)
else {
LOG(FATAL)
}

File: src/common/expression/PredicateExpression.cpp [hunk 2]
```diff
@@ -17,10 +17,18 @@ std::unordered_map<std::string, PredicateExpression::Type> PredicateExpression::
 };
 
 const Value& PredicateExpression::evalExists(ExpressionContext& ctx) {
-    DCHECK(collection_->kind() == Expression::Kind::kAttribute);
-    auto* attributeExpr = static_cast<AttributeExpression* >(collection_.get());
-    auto& container = attributeExpr->left()->eval(ctx);
-    auto& key = attributeExpr->right()->eval(ctx);
+    DCHECK(collection_->kind() == Expression::Kind::kAttribute ||
```
Comment (Shylock-Hg) [inline src/common/expression/PredicateExpression.cpp:20]: And this could remove.

**Review by nevermore3 — APPROVED** [review]

File: src/common/expression/PredicateExpression.cpp (hunk 1 above)
Comment (Aiee) [reply src/common/expression/PredicateExpression.cpp:26]: The expression type has been checked in the parser:
```
KW_EXISTS L_PAREN expression R_PAREN {
        if ($3->kind() != Expression::Kind::kLabelAttribute && $3->kind() != Expression::Kind::kAttribute &&
            $3->kind() != Expression::Kind::kSubscript) {
            throw nebula::GraphParser::syntax_error(@3, "The exists only accept LabelAttribe OR Attribute");
        }
        $$ = new PredicateExpression(new std::string("exists"), nullptr, $3, nullptr);
```
Is it still necessary to use the if-else here?

Comment (Shylock-Hg) [reply src/common/expression/PredicateExpression.cpp:26]: There is if-else already.

Comment (Aiee) [reply src/common/expression/PredicateExpression.cpp:26]: The `LOG(FATAL)` will never be printed because of the check from the parser tho.

Comment (czpmango) [reply src/common/expression/PredicateExpression.cpp:26]: IMO, if-else block is not necessary if you only need to get its operands.

**Review by nevermore3 — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Review by jievince — APPROVED** [review]

**Pull request merged**